### DISCUSSION TITLE
Repeters WIDE2-2  in frame APRS send position?

### DISCUSSION THREAD

**Discussion by F6DFA** [jgromes/RadioLib#1285] [Q&A]
In first time excuse me for my bad english
I try send a frame APRS with "WIDE2-2" so i have not in github the exemple an function of this in RADIOLIB.
the simple frame without "WIDE" is correct but
I try:
setRepeaters() an many X25frame:: ....  whithout result.
Thank's if you have solution of my problem

**Comment by jgromes** [comment]
I don't understand what exactly you're trying to achieve and what you tried, posting the code would help with that.

**Reply by F6DFA** [reply]
Thank's for answer
I make te detector of signal RF with HELTEC card V2 in used SX1276 . The modulation is FSK sample with APRS transmission for the data.
The frame APRS is correct but it'nt repete for relay because inside the frame position haven't the information WIDE2-2 
I try different writing 👍 whithout results

  //AX25Frame::AX25Frame(const AX25Frame& frame)
  //AX25Frame("APFD37",0,"WIDE",2,2);
  AX25Frame::setRepeaters("WIDE",'2',2);
  //ax25.setRepeaters(WIDE,repeteur,4);
   /*if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F(" X25 réussi !"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true) { delay(100); }
  }
  */
and on the github of Radiolib ,the haven't example and explication.
Please help me for finish my project
and excuse me for my bad english
PS: I have not habitude of the forum used and for transmit my code i put inside the text

**Reply by jgromes** [reply]
You seem to be mixing together multiple different things. RadioLib has documentation for every single method, including  [AX25Frame::setRepeaters](https://jgromes.github.io/RadioLib/class_a_x25_frame.html#a7f2d9f4f1ba29d0fd9f9f3f2cf03f797). In that documentation, you can see that:

1. It is part of the `AX25Frame` class, so you need to call it with an instance of that class.
2. It expects an array of repeater callsigns and SSIDs, because an AX.25 frame can contain up to 8 repeaters. If you want just one repeater, then both will be arrays of length one.

The following should work, assuming `AX25Frame frameUI` from the AX25_Frames example.

```c++
char* repeaterCallsigns[] = { "WIDE2" };
uint8_t repeaterSSIDs[] = { 2 };
state = frameUI.setRepeaters(repeaterCallsigns, repeaterSSIDs, 1);
```

**Reply by F6DFA** [reply]
thank's for information but at the compil i have:
  state = frameUI.setRepeaters(repeaterCallsigns, repeaterSSIDs, 1);
the warning:        _Compilation error: 'frameUI' was not declared in this scope_
in change the begining 👍 
   AX25Frame frameUI.setRepeaters(repeaterCallsigns, repeaterSSIDs, 1);
the warning:        Compilation error: APRS.h: No such file or directory
At this level , ma declaration is  only: 
#include <RadioLib.h>  // version 7.0.2 pour les différentes modulation dispo
I think that i don't write correctly the beginning of the sentense I don't anderstand.
what do you thing?

**Reply by jgromes** [reply]
Read my post again:

> The following should work, assuming AX25Frame frameUI from the AX25_Frames example.

`setRepeaters` is a method of the AX25Frame class. So you need to call it on an existing instance of that class. If you don't know what a class or its instance is, why are you trying to use a C++ library?

Here's the relevant section from the AX25_Frames example, with added repeaters.

```c++
  // create AX.25 Unnumbered Information frame
  // destination station callsign:     "NJ7P"
  // destination station SSID:         0
  // source station callsign:          "N7LEM"
  // source station SSID:              0
  // control field:                    UI, P/F not used, unnumbered frame
  // protocol identifier:              no layer 3 protocol implemented
  // information field:                "Hello World!"
  AX25Frame frameUI("NJ7P", 0, "N7LEM", 0, RADIOLIB_AX25_CONTROL_U_UNNUMBERED_INFORMATION |
                    RADIOLIB_AX25_CONTROL_POLL_FINAL_DISABLED | RADIOLIB_AX25_CONTROL_UNNUMBERED_FRAME,
                    RADIOLIB_AX25_PID_NO_LAYER_3, "Hello World (unnumbered)!");
  
  char* repeaterCallsigns[] = { "WIDE2" };
  uint8_t repeaterSSIDs[] = { 2 };
  frameUI.setRepeaters(repeaterCallsigns, repeaterSSIDs, 1);

  // send the frame
  Serial.print(F("[AX.25] Sending UI frame ... "));
  int state = ax25.sendFrame(&frameUI);
  if (state == RADIOLIB_ERR_NONE) {
    // the packet was successfully transmitted
    Serial.println(F("success!"));

  } else {
    // some error occurred
    Serial.print(F("failed, code "));
    Serial.println(state);

  }
```

**Comment by F6DFA** [comment]
Hello,
Thank's for the last message. I try insert the elements and have no problem for compilation.
So my frame position APRS is not repeted by the node but the frame UI is repeted.

i try many configuration:
1)
  AX25Frame frameUI("NJ7P", 0, "N7LEM", 0, RADIOLIB_AX25_CONTROL_U_UNNUMBERED_INFORMATION |
                    RADIOLIB_AX25_CONTROL_POLL_FINAL_DISABLED | RADIOLIB_AX25_CONTROL_UNNUMBERED_FRAME,
                    RADIOLIB_AX25_PID_NO_LAYER_3, "Hello World (unnumbered)!");
  
  char* repeaterCallsigns[] = { "WIDE2" };
  uint8_t repeaterSSIDs[] = { 2 };
  frameUI.setRepeaters(repeaterCallsigns, repeaterSSIDs, 1);
His position just after the init X25 (SETUP) , i thinks that is a configuration for the rest of programme;
result: Frame UI ok for repeted by the node but not the APRS frame position

2)
 AX25Frame frameUI("NJ7P", 0, "N7LEM", 0, RADIOLIB_AX25_CONTROL_U_UNNUMBERED_INFORMATION |
                    RADIOLIB_AX25_CONTROL_POLL_FINAL_DISABLED | RADIOLIB_AX25_CONTROL_UNNUMBERED_FRAME,
                    RADIOLIB_AX25_PID_NO_LAYER_3, "Hello World (unnumbered)!");
  
  char* repeaterCallsigns[] = { "WIDE2" };
  uint8_t repeaterSSIDs[] = { 2 };
  frameUI.setRepeaters(repeaterCallsigns, repeaterSSIDs, 1);
his position on the founction  "TX_APRS()" in" LOOP()"
result: the same résult frame UI OK - frame APRS NOK

The code tested follow:

     String infoUI = Gdh + Latitude + Longitude + TEXTE ;
     char TEXTEUI[100] = "";
     infoUI.toCharArray(TEXTEUI,100);

   AX25Frame frameUI("APFD37", 0, call, SSID, RADIOLIB_AX25_CONTROL_U_UNNUMBERED_INFORMATION |
                    RADIOLIB_AX25_CONTROL_POLL_FINAL_DISABLED | RADIOLIB_AX25_CONTROL_UNNUMBERED_FRAME,
                   RADIOLIB_AX25_PID_NO_LAYER_3, TEXTEUI);
                   
                   //RADIOLIB_AX25_PID_NO_LAYER_3, LATITUDE);//ok
                   // RADIOLIB_AX25_PID_NO_LAYER_3, "                            TEST DEV (unnumbered)!");//ok
                    //RADIOLIB_AX25_PID_NO_LAYER_3, GDH);//ok
                 

  char* repeaterCallsigns[] = { "WIDE2" };
  uint8_t repeaterSSIDs[] = { 2 };

  frameUI.setRepeaters(repeaterCallsigns, repeaterSSIDs, 1);
   // send the frame X25UI


  Serial.print(F("[AX.25] Sending UI frame ... "));
  state = ax25.sendFrame(&frameUI);
  if (state == RADIOLIB_ERR_NONE) {
    // the packet was successfully transmitted
    Serial.println(F("success!"));
  } else {
    // some error occurred
    Serial.print(F("failed, code "));
    Serial.println(state);}

 /* */  state = aprs.sendPosition("APFD37", 0, LATITUDE, LONGITUDE,TEXTE,GDH);    // lib radiolib
  if(state == RADIOLIB_ERR_NONE) {
          Serial.print(F(" TRAME APRS ENVOYEE AVEC succes! :"));Serial.println();
  } else {
          Serial.print(F("failed, code "));Serial.println(state);}

the frame APRS don't add "WIDE" at his frame - I dn't anderstand.
perhaps put any sign between information?

**Reply by jgromes** [reply]
OK, now I understand - what you're trying to do is to add repeaters to the APRS frame, not a generic AX.25 frame. That functionality was indeed missing in the library. I have added it in fada24c1c4e4cc65300309604506a209fbe652e3, the APRS_Position example is also updated now. I have tested it and it seems to be working - note the "Via WIDE2-2" in the second packet:

![Screenshot_116](https://github.com/user-attachments/assets/0ba648bb-ab84-48df-a8d0-40bcf6d5cd77)

**Comment by F6DFA** [comment]
ok Jan Thank's for new info
i try the modification on my code but 
i make releese radiolib before
thank's for your patience
alan

**Comment by F6DFA** [comment]
Hi jan
I have change the 4 files of librairy. I have no mistake and the frame APRS write WIDE but the node don't repete the frame
my configuration hard is in local
1 card ESP32 HELTEC V2 with GPS (F6DFA-11)
1 station Node PLX with UV5R (500mW) at 5m of ESP32 (F5ZLL-3)
1 station Visual PC with SARTRACK on BF888 (F6DFA-15)
1 node PLX distance 5km (5W) (F5ZLE-3)
The frame on the network APRS (F5ZLE-3) repete on my local node (F5ZLL-3)
the frame of my visu (F6DFA-15) also
The ESP (F6DFA-11)  is reveived on the node but no repete

The frame APRS is correct (see extract log)
:37/654] DATA: [23-10 22:04:37][AGWE-IN]  1:Fm F5ZLE-3 To APMI03 Via F5ZLL-3*,WIDE5-4 <UI pid=F0 Len=58 >[22:04:37]=4710.70N/00016.85E#                            N1 U=13.1V
[22:07:54/678] DATA: [23-10 22:07:54][AGWE-IN]  1:Fm F5ZLL-3 To APMI03 Via WIDE5-5 <UI pid=F0 Len=44 >[22:07:54]=4724.25N/00042.83E#     N0          U=10.8V
[22:07:56/700] DATA: [23-10 22:07:56][AGWE-IN]  1:Fm F5ZLL-3 To APMI03 Via F5ZLE-3*,WIDE5-4 <UI pid=F0 Len=44 >[22:07:55]=4724.25N/00042.83E#     N0          U=10.8V
[22:08:13/722] DATA: [23-10 22:08:13][AGWE-IN]  1:Fm **F6DFA-11 To APFD37 Via WIDE5-5** <UI pid=F0 Len=59 >[22:08:12]@200811z4708.62N/00015.67E>   S 99   Bat: 3.35   PARASITE !
[22:08:24/742] DATA: [23-10 22:08:24][AGWE-IN]  1:Fm F6DFA-11 To APFD37 Via WIDE5-5 <UI pid=F0 Len=59 >[22:08:24]@200818z4708.62N/00015.66E>   S 99   Bat: 3.35   PARASITE !
I think that the level X25 must actived but i am not specialist !

Nota;
My LCD is default when i change the file APRS (perhaps this my fault , i see to morrrow morning.

**Reply by F6DFA** [reply]
Hello Jan
i confirm the result:
i make the release of librairy this morning (github) replaced the last librairy
no mistake on compilation
same result:
no duplication of the frame APRS by the node 
I think that is the modulation is poor on AFSK  I change the different parameter in FSK but no amelioration !

**Reply by jgromes** [reply]
Well as far as I can see the repeater callsign and SSID is added correctly, so at that point I don't think it's an issue in the library.

> I think that the level X25 must actived but i am not specialist !
> I think that is the modulation is poor on AFSK I change the different parameter in FSK

I don't understand what you mean by that.

**Comment by F6DFA** [comment]
Hello Jan
excuse me for comback, Now i haven't problem for inside "WIDE" in APRS frame but the node APRS don't repete the frame because the modulation of ESP is bad.
I put a strapp betwen 25 and 34 (DO02) on my card HELTEC V2.
On 25 there are the square signal shift with modulation but no modulation 1200/2200Hz sinus.
on HF the audio is not good hear but is decoded on SARTRACK but not on SOUNDMODEM.
on the forum, i see that tone() is not supported on ESP32 (message 2020) and in radiolib i see use this fonction 
is it correct now ? 
Pehaps my déclaration is not correct, Please look on this under for on transmission on hamradio.
Thank's 


//-----------------------------------------
//------------ZONE- BIBLIOTHEQUES ---------
//-----------------------------------------
#include <arduino.h>      // APRS

// port COM carte ESP :    pas besoin de librairie specifique

// afficheur OLED interne carte HELTEC V2
#include <ESP32_LoRaWAN.h>

// afficheur LCD externe (gestion I2C)
#include <LiquidCrystal_I2C.h>  

//  GPS   -   cablé sur liaison Serie2
#include <TinyGPSPlus.h>
#include <SoftwareSerial.h>

// fonction RF 2.4GHz -  heltec dans bibliothèque declaree OLED
    // bluetooth
#include "BluetoothSerial.h"
#if !defined(CONFIG_BT_ENABLED) || !defined(CONFIG_BLUEDROID_ENABLED)
#error Bluetooth is not enabled! Please run `make menuconfig` to and enable it
#endif

    // WIFI  à développer utilisation en liaison serie
#include <WiFi.h>

// module LORA (150,400,868 MHz sur fiche SMA carte)
#include <SPI.h>        // pour gestion de la liaison serie et registres SX
#include <RadioLib.h>   // version 7.0.2 pour les différentes modulations disponibles
                        // prend en compte l'ajout modifié de la prise en compte 
                        // du WIDE en APRS


// BLUETOOTH
boolean bluetooth = false;    // variable 'etat liaison BT' ici 'coupée'
BluetoothSerial SerialBT;   // variable liaison série Bluetooth

//  WIFI
// à traiter

// LORA (module pour radio)
 // Broche LORA (initial TTGO LORA32 V1.0)  valable HELTEC V2 SX127x (base 868)
#define SCK 5       //    GPIO5  - SX1276's SCK  ok
#define MISO 19     //    GPIO19 - SX1276's MISO ok
#define MOSI 27     //    GPIO27 - SX1276's MOSI  ok
#define SS 18       //    GPIO18 - SX1276's Chip Select ok
#define pinRST 14   //    GPIO14   RESET  ok
#define pinIRQ 26   //    GPIO26   DI00 interrupt request
#define DI01 35     //    GPIO35 - DI01 SX1276  ajout pour radiolib
#define DI02 34     //    GPIO35 - DI02 SX1276  ajout pour radiolib

SX1278 radio1 = new Module(18, 26, 14, 35);  // radiolib
// variables et constantes

AFSKClient audio(&radio1, 25);  // generation des tones sur pin 25 de l'ESP heltec V2(strapp ou resistance vers pin 34 DI02)
MorseClient morse(&audio);
PagerClient pager(&radio1);
//RTTYClient rtty(&audio);// en DEFAUT
AX25Client ax25(&audio);
// create APRS client instance using the AX.25 client
APRSClient aprs(&ax25);
//BellClient bell(&radio1, 25);

SETUP
// affectation du sens des pins (si imposé)
  pinMode(led_intern, OUTPUT);            // led carte
  pinMode(pinACQ, INPUT_PULLUP);          // bouton poussoir ACK
  pinMode(pinBAT, INPUT_PULLUP);          // entree mesure de la tension batterie
  pinMode(pinDEV, INPUT_PULLUP);          // entree mesure et controle Developpement

  pinMode(SCK, OUTPUT);                   // lora
  pinMode(MOSI, OUTPUT);                  // lora
  pinMode(MISO, INPUT_PULLUP);            // lora
  pinMode(SS, OUTPUT);                    // Chip select lora
  pinMode(pinRST, OUTPUT);                // reset
  pinMode(pinIRQ, INPUT_PULLUP);          // lora DI00
  pinMode(DI01, OUTPUT);                  // lora DI01
  pinMode(DI02, INPUT_PULLUP);            // lora DI02 
  
// Initialisation vitesse port COM à 115200 bauds
  Serial.begin(115200);

// Initialiser interface de l'OLED
    //RAZ de l'afficheur OLED = reset bas pin GPIO16 (sinon ne fonctionne pas)                    
    // impulsion négative de 50ms
  pinMode(16,OUTPUT);
  digitalWrite(16, LOW); // set  GPIO16 low to reset OLED intern
  delay(50); 
  digitalWrite(16, HIGH); // while OLED is running, must set GPIO16 to high
   
  oled.init(); // SSD 1306
   //Inverser le sens d'affichage de l'oled si besoin. 
   // ATTENTION  :   Mettre les 2 instructions suivantes sinon problème.
   // oled.flipScreenVertically();
   // oled.setFont(ArialMT_Plain_10);
  
// Initialiser interface du LCD (choisir modele)
  lcd.begin(16,2,0); // LCD 16 colonnes 2 lignes taille 8*4 pixel caractere=0
  //lcd.begin(20,4,0); // LCD 20 colonnes 4 lignes taille 8*4 pixel caractere=0

// initialisation LED bleu interne
  digitalWrite(led_intern, LOW); // led allumé mais prépositionné par init lora afsk ?
    
// Initialisation du GPS -  
  ss.begin(vitesse_GPS);        //vitesse liaison serie du GPS ublox (9600)
                                //vitesse liaison serie du GPS GOSGET BU358 (4800)
// Initialisation BLUETOOTH
  SerialBT.begin("SIT406_detecteur"); //Bluetooth device name
// Initialisation module LORA interne
   // initialize FSK SX1276  with default settings------------
  Serial.print(F("[SX1278] Initialisation FSK ... "));
  // when using one of the non-LoRa modules for AFSK
  // (RF69, CC1101, Si4432 etc.), use the basic begin() method
  // int state = radio.begin();
//int  state = radio1.beginFSK(frequenceRX);
//int  state = radio1.beginFSK(frequenceRX,vitesse_AFSK);
int state = radio1.beginFSK(frequenceRX,vitesse_AFSK,5.0 ,125.0 ,10 ,8 ,false);
//int state = radio1.begin(frequenceRX,vitesse_AFSK,5.0 ,125.0 ,10 ,8 ,false);
  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F(" INIT FSK réussie!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true) { delay(100); }
  }
 
/**/
// initialize AFSK client----------------------------              ok (strap 25-34)sur heltec v2
  Serial.print(F("[AFSK] Initialisation ... "));    // genere tonalité sur pin DAC 25
   state = audio.begin();
  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F(" AFSK réussie!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true) { delay(100); }
  }
   // initialize AX.25 client
  Serial.print(F("[AX.25] Initialisation ... "));
  // source station callsign:     "N7LEM"
  // source station SSID:         0
  // preamble length:             8 bytes
  state = ax25.begin(call,SSID);                             
  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F(" X25 réussi !"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true) { delay(100); }
  }
  
  // initialize APRS client
  Serial.print(F("[APRS] Initialisation ... "));
  // symbol:                      '>' (car)
  state = aprs.begin('>');
  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F("APRS réussi !"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true) { delay(100); }
  }

**Comment by F6DFA** [comment]
hi jan
my result with modification preamble = 32 on X25
1:Fm F6DFA-11 To APFD37 Via F5ZLL-3*,WIDE5-4 <UI R Pid=F0 Len=86> [16:14:12R] [+++]
@141408z4708.62N/00015.66E>14 H 14 mn 8 N47.1437E0.2610  S 99   Bat: 3.35 trame COURTE
2:Fm F6DFA-11 To APFD37 Via F5ZLL-3*,WIDE5-4 <UI R Pid=F0 Len=86> [16:14:12R] [+++]
@141408z4708.62N/00015.66E>14 H 14 mn 8 N47.1437E0.2610  S 99   Bat: 3.35 trame COURTE

F6DFA-11 : ESP32 card
F5ZLL-3 : node local (5meters)
on 1 Hour with the beacon every 5mn, this is 10% of reply 
have posibility of TXdelay parameter  and perhaps is't possible to put the frequency (sinus) inside AFSK?

**Comment by jgromes** [comment]
> On 25 there are the square signal shift with modulation but no modulation 1200/2200Hz sinus.

You're not going to see a sine wave there. The radio is digital, it cannot perform a true AFSK modulation. It can approximate it by using 1200 Hz and 2200 Hz square wave, which is what RadioLib generates.

> on the forum, i see that tone() is not supported on ESP32 (message 2020) and in radiolib i see use this fonction
is it correct now ?

RadioLib uses PWM to emulate Arduino tone function on ESP32. But that's not the point, tone() cannot produce sine wave on most Arduino platforms.

> on 1 Hour with the beacon every 5mn, this is 10% of reply
> have posibility of TXdelay parameter and perhaps is't possible to put the frequency (sinus) inside AFSK?

I don't understand what you're trying to say. I feel like the language barrier here is too large.

**Reply by F6DFA** [reply]
thanks for the information on the tone generator in ESP,
I wanted to insert a longer delay before transmitting the frame (TXDELAY) because the TXs do not all have the same delay upon receipt of the TX command
regarding the lengthening of the preamble from 8 to 32, the reliability of the node's understanding can be due to several factors:
the fallout of the RX squelch 
the rise of the sqelch which alters the received frame
this is valid for remote reception in particular as the transmission duration is very short in comparison with other APRS signals

**Reply by jgromes** [reply]
> I wanted to insert a longer delay before transmitting the frame (TXDELAY) because the TXs do not all have the same delay upon receipt of the TX command

Where is this delay exactly? If it's between subsequent APRS transmissions, then that's completely up to you in the ESP32 code. Or is it somewhere in the AX.25 frame itself ...?

Regarding the preamble length, you can try even longer ones, I don't know what the comon values are, but there should be no limit of the preamble length in RadioLib.

**Comment by F6DFA** [comment]
hello Jan
my model is made up of a PLX node which controls a UV5R talky (this is the configuration of our APRS network which works with the Arduino detector). 
a SARTRACK supervision console is made with a UV5R, it decodes ESP frames without problem
on the PLX node the frame sent is well signaled (LED) and well decoded on the supervision console
my reasoning is that the time between the time the ESP goes into transmission and the sending of the frame is too short to overcome the rise in TX of the transceivers
Radiolib does not allow you to modify this time. 
Concerning the preamble I raised it up to 255 without result (modified on FSK and AX25)
as noted in a 2020 forum post.
what do you think?

**Reply by F6DFA** [reply]
I just did preamble tests again 
a value between 72 and 148 on X25 initialization seems more reliable for recovery by a node. outside of these values ​​<72 the success rate is low, around 5 to 10%, above 155 I have an ESP reset
I kept the preamble of 8 on the FSK init

**Reply by jgromes** [reply]
> my reasoning is that the time between the time the ESP goes into transmission and the sending of the frame is too short to overcome the rise in TX of the transceivers

That's what the preamble should be for. It doesn't carry any information, it just lets the receiver know that some transmission will happen. It doesn't matter if some preamble bytes  get lost. So I don't think there's any need to add extra delay.

**Comment by F6DFA** [comment]
hello Jan
Thank you for your information which helps me move forward with this detector.
I have tried several pieces of equipment which do not all react with the same number of preambles, looking more closely, the characteristics of the equipment may be the cause (I have 2 which have a poorly adjusted disc!)
Anglais

currently I have a node which takes frames at 100% with a BAOFENG BF888 and a preamble at 98
so I think that the development can stop here to devote myself to another area to develop, phase decoding from the SX1278 to bring out the decoding of the frame.
Thank you again for your patience and your availability
Best friendships
alain
